controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -565,6 +806,21 @@ func (pp *portPublisher) updatePort(targetPort namedPort) {
 	}
 }
 
+func (pp *portPublisher) deleteEndpointSlice(es *dv1beta1.EndpointSlice) {
+	addrSet := pp.endpointSliceToAddresses(es)
+	for id := range pp.addresses.Addresses {
+		if _, found := addrSet.Addresses[id]; found {
+			delete(pp.addresses.Addresses, id)
+		}
+	}
+
+	if len(pp.addresses.Addresses) > 0 {
+		return
+	}
+
+	pp.noEndpoints(false)
```

Comment:
Hm, is this here correct? If a slice is deleted, you are calling noEndpoints(false). Correct me if I am wrong but this should send an update saying that the svc_name:port does not exist anymore. If you delete an endpoint slice does that mean that we should be sending exists=false? @adleong wdyt ? Also, does that really mean that no Endpoints need to be send. There might be other endpoints in other slices. I have the feeling I might be getting all this wrong. Have to think about that a bit.

Reply:
So from the comments of the proxy api: 

```
// If the service does not exist then the controller must send
// `no_endpoints{exists: false}` ASAP when a client subscribes or when the
// service stops existing. If the service exists and has endpoints available
// then the controller must send `add` that lists all (or at least a large
// number) of the endpoints for the service. If and only if the service exists
// but does not have any endpoints available then the controller SHOULD send
// `no_endpoints{exists: true}` when a client subscribes. In other words, the
// `no_endpoints` message must only be sent when there are *no*endpoints for
// the service.
```

Reply:
Yes I think there are two problems here: 

- we cannot inform the proxy that there are no endpoints simply because a slice was deleted because there might be other slices that are associated with this service
- we certainly cannot inform the proxy that this service has no associated endpoints because of the above as well.

Reply:
Thanks for the detailed comment @zaharidichev it all makes sense! I was a bit confused around the `NoEndpoints` concept.

I wanted to send a `NoEndpoints` message only if the port publisher has no addresses registered anymore, however. The way I thought this was:
* becase a svc:port might have loads of addresses split across different slices, go through the current addresses for that port publisher and remove them.
* if at the end we do not have any addresses, send a no endpoints message.

I tried to do this with the line above:
```
if len(pp.addresses.Addresses) > 0 -- return out of the function
```

perhaps this is not the most intuitive way to do though, any thoughts?

Reply:
I think it will be clearer if you do something like: 

```
svcExists := len(pp.addresses.Addresses) > 0
pp.noEndpoints(svcExists)
```

Reply:
But all that brings an interesting for me question. Namely does the fact that a service has no endpoints (and hence noe addresses) mean that it should be non existent for the purposes of discovery? I see that in `deleteEndpoints` we used to fire a NoEndpoints(false).  I guess this was designed like that for a reason. Coming from some previous discussions around services that do not have any endpoints and are there just as proxies (for mc purrposes), I think it is a valid scenario to do a discovery on a service that does not have any endpoints.. or is it not? @adleong wdyt? 

@Matei207  Note, this is not directly related to this PR. I am just thinking out loud. Your approach is correct for the time being.

Reply:
> I think it will be clearer if you do something like (...)

That's an amazing suggestion, thank you! It reads much better.

> But all that brings an interesting for me question (...) does the fact that a service has no endpoints (and hence no addresses) mean that it should be non-existent for the purposes of discovery?

No, at least I don't think so. This is a very good question, based on the test cases and the general functionality that is provided by the destination service, I think in the case of a service that does not have any endpoints available, listeners should still be able to subscribe to the service and receive a "NoEndpoints" message. Since on subscription the client (proxy) receives a stream of updates, it's only fair that a service be available for discovery even if no endpoints (addresses) exist; my assumption is that there might be a stream of updates propagated to the client at some point in the future (?), pods can come up (and disappear) at any time.

So @zaharidichev I agree with that it is valid to do a discovery on a service that does not have any endpoints, based on what I've seen so far.